pkg/sql/sem/tree/create_logical_replication.go

```diff
@@ -0,0 +1,174 @@
+		ctx.WriteString(" WITH ")
+		ctx.FormatNode(&node.Options)
+	}
+}
+
+func (lrr *LogicalReplicationResources) Format(ctx *FmtCtx) {
+	if lrr.Database != "" {
+		ctx.WriteString("DATABASE ")
+		lrr.Database.Format(ctx)
+	} else if len(lrr.Tables) > 1 {
+		ctx.WriteString("TABLES (")
+		for i := range lrr.Tables {
+			if i > 0 {
+				ctx.WriteString(", ")
+			}
+			ctx.FormatNode(lrr.Tables[i])
+		}
+		ctx.WriteString(")")
+	} else {
+		ctx.WriteString("Table ")
+		ctx.FormatNode(lrr.Tables[0])
+	}
+}
+
+func (lro *LogicalReplicationOptions) Format(ctx *FmtCtx) {
+	var addSep bool
+	maybeAddSep := func() {
+		if addSep {
+			ctx.WriteString(", ")
+		}
+		addSep = true
+	}
+
+	if lro.Cursor != nil {
+		maybeAddSep()
+		ctx.WriteString("CURSOR = ")
+		ctx.FormatNode(lro.Cursor)
+	}
+
+	if lro.DefaultFunction != nil {
+		maybeAddSep()
+		ctx.WriteString("DEFAULT FUNCTION = ")
+		ctx.FormatNode(lro.DefaultFunction)
+	}
+
+	if lro.Mode != nil {
+		maybeAddSep()
+		ctx.WriteString("MODE = ")
+		ctx.FormatNode(lro.Mode)
+	}
+
+	if lro.UserFunctions != nil {
+		maybeAddSep()
+		addSep = false
+
+		// In order to make tests deterministic, the ordering of map keys
+		// needs to be the same each time.
+		keys := make([]TablePattern, 0, len(lro.UserFunctions))
+		for k := range lro.UserFunctions {
+			keys = append(keys, k)
+		}
+		sort.Slice(keys, func(i, j int) bool {
+			return keys[i].String() < keys[j].String()
+		})
+
+		for _, k := range keys {
+			maybeAddSep()
+			ctx.WriteString("FUNCTION ")
+			ctx.FormatNode(lro.UserFunctions[k])
+			ctx.WriteString(" FOR TABLE ")
+			ctx.FormatNode(k)
+		}
+	}
+}
+
+func (o *LogicalReplicationOptions) CombineWith(other *LogicalReplicationOptions) error {
+	if o.Cursor != nil {
+		if other.Cursor != nil {
+			return errors.New("CURSOR option specified multiple times")
+		}
+	} else {
+		o.Cursor = other.Cursor
+	}
+
+	if o.Mode != nil {
+		if other.Mode != nil {
+			return errors.New("MODE option specified multiple times")
+		}
+	} else {
+		o.Mode = other.Mode
+	}
+
+	if o.DefaultFunction != nil {
+		if other.DefaultFunction != nil {
+			return errors.New("DEFAULT FUNCTION option specified multiple times")
+		}
+	} else {
+		o.DefaultFunction = other.DefaultFunction
+	}
+
+	if other.UserFunctions != nil {
+		for tbl := range other.UserFunctions {
+			if _, ok := o.UserFunctions[tbl]; ok {
+				return errors.Newf("multiple user functions specified for table %q", tbl)
+			}
+			if o.UserFunctions == nil {
+				o.UserFunctions = make(map[TablePattern]Expr)
+			}
+			o.UserFunctions[tbl] = other.UserFunctions[tbl]
+		}
+	}
+
+	return nil
+}
+
+// IsDefault returns true if this logical options struct has default value.
+func (o LogicalReplicationOptions) IsDefault() bool {
+	options := LogicalReplicationOptions{}
+	return o.Cursor == options.Cursor &&
+		o.Mode == options.Mode &&
+		o.DefaultFunction == options.DefaultFunction &&
+		(o.UserFunctions == nil && options.UserFunctions == nil)
```

Comment:
I'm not sure we need options.UserFunctions == nil. If you wanted to follow the pattern of the others, comparing to the empty opts structs zero values instead of checking zero values directly, I'd say just len compare these to eachother. Otherwise I think just the first side is enough on its own